BUG,MAINT: Fix issues with non-reduce broadcasting axes

seberg opened the pull request:
There are cases where we want to be OK with broadcasting axes (basically making them reduction axes) and some cases where we are not. For normal reductions when an output array is specified, it is possible that the output array has non-reduction axes with mismatching shape (but length 1). These axes should not be interpreted as reduction axes.
On the other hand, in non-reductions (e.g. generalized ufuncs), "reduction axis" are added to operands, but they are deleted later on. This also means that for such axes, the shape must not be checked. (In other words "Reduction Axis" in this sense actually means: "As long as the shape is 1, do not worry if this axis is a reduction or not).

Up for now for discussion with @mhvk, although with some tweaks it may be nice. For better understanding:
  * `-1` means that this operand has an (implicit) `np.newaxis` in the full iteration shape.
  * `-1` axis are probably only ever added when we are OK with them being broadcast (either because it is reduction, or because we will remove them later); However, that is not guaranteed?
  * A positive axis, may or may not be a broadcast axis. With `keepdims=True` it is not a new (`-1`) dimension. If it is a broadcast one, we have to set its length to `1` (and check that it is one).

I would need to carefully go through to see if there is no whole in the logic during the NpyIter construction. The current use of the `NPY_ITER_REDUCE_OK` as a "flag" is not a good final solution, maybe two macros to handle this could be, and we have to see whether we make it public API (right now everything is, since multiarray and umath were split).

EDIT:

One note on the reduction changes *when there is no identity*: Previously if the reduction had a single axis, the input operand was shrunk by one before the iterator was set up. This simplifies the logic later on (probably slightly increasing the speed). There was also a fast path, to not do any iteration at all in that case if _no_ reduction axis was specified. That seems too corner case to be relevant to me.

EDIT: This is an alternative to gh-15142  and gh-15144.

mhvk commented on numpy/core/src/umath/reduction.c:
```diff
@@ -267,18 +45,15 @@ count_axes(int ndim, const npy_bool *axis_flags)
- * exception when there are no elements to reduce (which appropriate iff the
- * reduction operation has no identity).
+ * fully initialized to the identity, because this function raises an
+ * exception when there are no elements to reduce (which is appropriate iff
```
(existing) typo: iff->if

seberg replied:
Its "if and only if" :), IShould change it though, its easy to stumble over these and for most readers its just a typo.

mhvk commented on numpy/core/src/umath/reduction.c:
```diff
@@ -293,98 +68,86 @@ count_axes(int ndim, const npy_bool *axis_flags)
             ++nreduce_axes;
         }
+        else {
+            if (!keepdims) {
```
It's a total nitpick, but I would instead initialize a `idim_out=0` at the start, and increment it here and inside the `if (keepdims)` clause above (and then always set `shape[idim_out]` etc). (This of course is similar to what you do with `curr_axis` below - I find it more readable somehow.)

mhvk commented on numpy/core/src/umath/reduction.c:
```diff
@@ -293,98 +68,86 @@ count_axes(int ndim, const npy_bool *axis_flags)
-    if (PyArray_CopyInto(result, op_view) < 0) {
-        Py_DECREF(op_view);
-        return NULL;
+    if (!keepdims) {
```
With the above, this could become an unconditional `ndim = idim_out`. Though I think it is probably better to use a new name (just `idim_out`??)

mhvk commented on numpy/core/src/umath/reduction.c:
```diff
@@ -293,98 +68,86 @@ count_axes(int ndim, const npy_bool *axis_flags)
+ * Returns 0 on success and -1 if an error occurred.
  */
-NPY_NO_EXPORT PyArrayObject *
+NPY_NO_EXPORT int
```
Slightly bigger: how about making this a function that just returns a view of the first element? (i.e., returns `op_view`) It feels that the copy at the end doesn't really belong inside here, but rather should be done in the caller.

seberg replied:
Opted for renaming the function for now, not sure what is nicer.

mhvk replied:
Returning the size makes sense.

mhvk commented on numpy/core/src/umath/reduction.c:
```diff
@@ -537,15 +263,83 @@ PyUFunc_ReduceWrapper(PyArrayObject *operand, PyArrayObject *out,
+    for (int i = 0; i < PyArray_NDIM(operand); i++) {
+        if (axis_flags[i]) {
+            if (keepdims) {
+                result_axes[i] = curr_axis + NPY_ITER_REDUCE_OK;
```
This assignment is the one for which my suggestion of setting to `-2` would not work, right?

seberg replied:
Yeah, its where you need a flag for an existing axis. You could try to not use that (never use keepdims) but that does not seem like a right approach.

In your combination of numbers and flags, what do you mean exactly? Do you mean making `op_axes` longer by inserting flags into it? I do not quite see why newaxis is a flag in this regard.  (Remember this is a mapping `iter_axis→op_axis` and newaxis means it does not exist in the op.)

mhvk commented on numpy/core/src/umath/reduction.c:
```diff
@@ -537,15 +263,83 @@ PyUFunc_ReduceWrapper(PyArrayObject *operand, PyArrayObject *out,
+        }
+    }
+    else {
+        if (PyArray_InitializeReduceResult(
```
Seeing the use here, this is also quite neat. Still, maybe even clearer as
```
PyArray_Object *first_elements = first_elements_view(operand, axis_flags, funcname, keepdims);

if (first_elements == NULL) {
    goto fail;
}
if (PyArray_CopyInto(result, first_elements) < 0) {
    Py_DECREF(first_elements);
    goto fail;
}
Py_DECREF(first_elements);
```

mhvk reviewed (commented):
Some small comments, but really all those can safely be ignored. I do think this is a very substantial improvement, given its large simplification in `reduce.c`. The main question remains whether there isn't a better alternative implementation beyond adding the flag to signal reduction is OK...

Just for other reviewers, let me copy from https://github.com/numpy/numpy/pull/15142#issuecomment-568576924:

On the implementation: would it make more sense to re-interpret `op_axes` as a combination of axis number and flags?  With `NPY_ITER_REDUCE_OK` and `NPY_ITER_NEWAXIS` the two possible flags (and `-1` special-cased to mean just the latter for backwards compatibility). In principle, this could be exported on the python side by allowing `np.newaxis` inside the `op_axes` argument - though not as neatly for the reduction...

mhvk commented:
(Taking discussion of possible flags out of the in-line comments, since those become invisible)

I guess my suggestion was to start treating `op_axes` as, say, a union of a few-bits (16?) int for the axis number and other bits for flags (upper 16 bits?). This would do away with `-1` altogether (though it would need to be kept for backward compatibility) and instead have its meaning be taken by a new `NPY_ITER_NOAXIS` flag (you're right it is not really `NEWAXIS`, though it is described as such in the description of `nditer` and for inputs that makes some sense).

Though I must admit I'm far from sure this is actually better... maybe ping @eric-wieser - he usually has good ideas...

seberg commented:
Hmm, yeah. Which just means that the flag extraction is different for positive and (negative axes) newaxis. On the up-side maybe, we could get the axes (and -1) with a simple `(int16)axis` cast.

mhvk commented:
I thought to just have as the very first line a backward compatibility check along the lines of `if (axis == -1) { axis=NPY_ITER_NO_AXIS }`

seberg commented:
One other thing. If we go down this road, I am actually thinking we should maybe deprecate `NPY_ITER_REDUCE_OK` entirely. It does not really seem like a correct thing to do normally. There is one more place we use (in einsum) where it is trivial to fix. There it is not a bug I think, but only because `NPY_ITER_NO_BROADCAST` is used, and I suspect that also may not be needed if we get rid of the `NPY_ITER_REDUCE_OK` flag.

Flagging the axes for "Reduction" seems just cleaner. I am not quite sure if I am already enforcing that if "Reduction" is set for an axes, its length must be `1`.

mhvk commented:
Yes, not sure there is much point in a general flag for letting the iterator know that reduction is OK. That said, passing in `op_axes` is a bit of a pain.

seberg commented:
Yeah, one option might be to allow `NPY_ITFLAG_REDUCE_OK` but only on operands for which `*op_axes == NULL`. OTOH, it is a pain, but it is a mild pain maybe. If you did not allocate the array already, the iterator is going to currently not make it a reduction anyway, so there is no point.

If the array is user provided, you should probably be doing a size check anyway, so moving it to the iterator is OK. If you allocate the array currently, you can probably again move that to the iterator and actually reduce the code (like in reductions).

eric-wieser commented on numpy/core/src/multiarray/nditer_constr.c:
```diff
@@ -1410,6 +1413,59 @@ check_mask_for_writemasked_reduction(NpyIter *iter, int iop)
+ * should be reduced. If axes are specified explicitely, the flag is
+ * not needed
+ */
+static int
+npyiter_check_reduce_ok_and_set_flags(
+        NpyIter *iter, npy_uint32 flags, npyiter_opitflags *op_itflags) {
+    /* If it's writeable, this means a reduction */
+    if (*op_itflags & NPY_OP_ITFLAG_WRITE) {
+        if (!(flags & NPY_ITER_REDUCE_OK)) {
+            PyErr_SetString(PyExc_ValueError,
+                    "output operand requires a reduction, but reduction is"
+                    "not enabled");
+            return 0;
+        }
+        if (!(*op_itflags & NPY_OP_ITFLAG_READ)) {
+            PyErr_SetString(PyExc_ValueError,
+                    "output operand requires a reduction, but is flagged as "
+                    "write-only, not read-write");
+            return 0;
+        }
+        NPY_IT_DBG_PRINT("Iterator: Indicating that a reduction is"
+                         "occurring\n");
+
+        NIT_ITFLAGS(iter) |= NPY_ITFLAG_REDUCE;
+        *op_itflags |= NPY_OP_ITFLAG_REDUCE;
+    }
+    return 1;
+}
```
Would recommend returning `-1` for errors, not 0.
```suggestion
static int
npyiter_check_reduce_ok_and_set_flags(
        NpyIter *iter, npy_uint32 flags, npyiter_opitflags *op_itflags) {
    /* If it's writeable, this means a reduction */
    if (*op_itflags & NPY_OP_ITFLAG_WRITE) {
        if (!(flags & NPY_ITER_REDUCE_OK)) {
            PyErr_SetString(PyExc_ValueError,
                    "output operand requires a reduction, but reduction is"
                    "not enabled");
            return -1;
        }
        if (!(*op_itflags & NPY_OP_ITFLAG_READ)) {
            PyErr_SetString(PyExc_ValueError,
                    "output operand requires a reduction, but is flagged as "
                    "write-only, not read-write");
            return -1;
        }
        NPY_IT_DBG_PRINT("Iterator: Indicating that a reduction is"
                         "occurring\n");

        NIT_ITFLAGS(iter) |= NPY_ITFLAG_REDUCE;
        *op_itflags |= NPY_OP_ITFLAG_REDUCE;
    }
    return 0;
}
```

eric-wieser replied:
Hmm, I see the rest of this file does this. Perhaps a cleanup patch to switch the meaning would be sensible, but perhaps it would just be churn.

mhvk commented:
@seberg - any chance to wrap this up? Can I help?

seberg commented:
Hmmm, I have to look at it in more detail. If I remember right, I ran a bit too far down the try to clean-up/simplify the code more, but not sure how much I pushed or not...

In any case, if we can agree on such a flag meaning "may broadcast axis" (no matter if readonly/writeonly maybe) and the general flagging stuff. I think I can try to see if I can fix this up at the end of the week... Will be a bit annoying to review :(, and I guess that would be the biggest way you can help in the end...

mhvk commented:
@seberg - this PR solved a real issue and indeed I think we got drawn too far to trying to have a perfect solution rather than one that is good enough. My sense is to go essentially for the solution you had, and worry about beauty if this also gets exposed on the python side. Anyway, bottom line: happy to review. Even if it is just a rebase of what you have here (which I mostly reviewed, but there is a bit of new stuff).

seberg commented on numpy/core/include/numpy/ndarraytypes.h:
```diff
@@ -1075,6 +1075,9 @@ typedef void (NpyIter_GetMultiIndexFunc)(NpyIter *iter,
 #define NPY_ITER_PER_OP_FLAGS               0xffff0000
 
 
+#define NPY_ITER_FORCED_BROADCAST_AXIS(axis) (axis + (NPY_MAX_INT >> 1))
```
This is a public header, we have to decide whether we think this is OK. Or move it to a private one for now. (That will be my default if nobody says anything. But to be clear. My feeling is that this is the much better API. As we can see based on all the weird behaviour, the general reduce OK flag does not really make much sense!)

seberg replied:
A good alternative may be to make this public as a function, not macro, so that it can be used but we are free to change how the flag looks like.

seberg replied:
Just to note, some things went wrong in the rebase probably... Or there were some bugs and I did not realize it... in any case some more fixup needed here.

mhvk replied:
Making it private makes sense.  Function is OK too, but I'd go for minimum work...

seberg replied:
Yeah, will just move it to a private header later... Right now I got something lost with the broadcast axis being length 1 and that being lost...

seberg replied:
Moved it to `common.h` for now, should be OK although does not feel right. I have used `defined(NPY_INTERNAL_BUILD)` before, but that does not really seem right either?

In any case, its moved. And I guess there was a silly mistake in the last commit. The last commit is some general code duplication reduction and I could remove it, or reorder it so that the diff is possibly a bit simpler.

Unfortunately the whole giant reduce cleanup did not seem to be easy to disentangle :/...

seberg replied:
(sorry for the noise, this rebasing is going wrong, I will have to finish it tomorrow... including finishing the renaming)

seberg committed:
MAINT: Simplify undoing of iter-axis-perm in nditer

seberg committed:
MAINT: (nditer) move duplicated reduce axis check into helper func

seberg committed:
MAINT: Simplify signature of npyiter_replace_axisdata

seberg committed:
MAINT: Simplify strides and axis check in npyiter_new_temp_array

The stride calculation and axis check redid some of the work
finding the number of dimensions again. This simplifies the logic
and clarifies an error message a bit.

seberg committed:
ENH: Allow specifying reduction axis in nditer/NpyIter

A reduction axis can be a broadcast axis (-1), which does not
exist. Or an existing axis which then must have a length of 1, or
will be allocated with a length of 1.

This allows the correct allocation of reduction outputs with
axes of size 1 (instead of only newaxis/-1), i.e. it allows
supporting the result output with `keepdims=True` in reductions.

It can also be used by generalized ufuncs to specific that one
operand does not need to iterate a core dimensions. In the case
of generalized ufuncs, this is handled like a reduce dimensions,
but that dimension is later removed. This means that the operand
can be specified without read-write.
It mainly is necessary to allow normal broadcasting for the axes
_not_ marked as reduce axis.

In most cases however, an operand should never use broadcasting
when marking axes as reduction.

seberg committed:
MAINT: Use reduction axis marker for einsum

seberg committed:
BUG: Avoid incorrect broadcasts on non-core outputs in gufuncs

Mark the core dimensions as reduce axis. This allows them to be
freely broadcast (basically ignored for most practical purposes)
while other axes are normally broadcast even though the operand is
read-only or write-only.
If we were to use read-write operands with `REDUCE_OK` these
additional dimensions are currently simply absored as reduction
dimensions...

(I hope this is understandable, please see the issues/test
for an example...)

Fixes gh-15139, replaces gh-15142

Co-Authored-By: Marten van Kerkwijk <mhvk@astro.utoronto.ca>

seberg committed:
BUG,MAINT: Simplify reduction result creation using reduce axis

The new reduce axis marker for nditer/NpyIter allows to move the
allocation logic necessary to allocate the reduction result
to the nditer, thus simplifying the logic here.

This leads to some other related cleanups of the result
initialization.
The subok flag was removed, since it was always set to 0
and the iterator rejects subclasses.

seberg committed:
DOC: Slightly improve error message on incorrectly shaped reductions

seberg commented:
@mhvk OK, I finally got things somewhat straightened out. I may look into adding python side tests by manually adding the flag.

The reduce change is unfortunately a giant chunk, since it deletes a lot of code to fix the reduction related issues (this makes the test originally added by you much more strict, which I think is the right thing though).

The are a couple of `MAINT` commits at the beginning, which I could split out, since they are really just attempts to to shorten those long functions a bit. The `Simplify strides and axis check in npyiter_new_temp_array` commit may be a bit trickier.

In general, this should be best reviewed by looking through the commit one by one. If its too large of a chunk, please tell me. I should self-review the reduction changes before the end...

anirudh2290 commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -610,7 +610,17 @@ def test_broadcast(self):
+        assert_raises(ValueError, umt.inner1d, a, b, out)
+        out2 = np.empty(3)
+        c = umt.inner1d(a, b, out2)
+        assert_(c is out2)
```
are all these tests supposed to work today. I am asking these because I tried and atleast a subset of them seem to be working on the master. I was looking for a test that breaks today, that was fixed by this PR.

seberg replied:
Hmmm, the error should not be raised on master? The second one is the correct version, hmm, I don't remember, it could be that it is well tested above but I guess there were few `out` tests?

seberg replied:
The nditer tests just excercise the code branch where I did some maintenance here (I am happy to roll those changes out if it helps get the rest in.)

anirudh2290 replied:
I mean  I tried the ones in test_reduce_wrong_dimension_output also and that passed too . which test should i try to see what fails today and fixed by this PR.

seberg replied:
```
from numpy.core import _umath_tests as umt
a = np.arange(6).reshape(3, 2)
b = np.ones(2)
out = np.empty(())
umt.inner1d(a, b, out)
```
does not raise an error?

anirudh2290 replied:
ya this doesnt. my version is 1.20.0.dev0+ba4e4b6 , also tried on 1.18.1

seberg replied:
woops, sorry, that is supposed to raise an error :). But, I am not immediatley sure about `out=np.empty((3,3))` or so (out being too large). I guess that should raise as well or broadcast the inputs to the output (which half makes sense), may be good to add a test for it either way.

mhvk replied:
These tests come from my PR at #15142 and were partially meant to increase the test coverage, i.e., include things that were already OK but not tested anywhere.

mhvk replied:
@seberg - I think the test is fine as is: a is 2-dimensional so the expected output shape is (3,)

anirudh2290 replied:
> woops, sorry, that is supposed to raise an error :). But, I am not immediatley sure about `out=np.empty((3,3))` or so (out being too large). I guess that should raise as well or broadcast the inputs to the output (which half makes sense), may be good to add a test for it either way.

i am not sure what the correct behavior is, but this doesnt seem to raise an error with this change or broadcast the inputs to the output.

```
import numpy as np
from numpy.core import _umath_tests as umt
from numpy.testing import assert_raises
a = np.arange(6).reshape(3, 2)
b = np.ones(2)
out = np.empty((5, 3))
umt.inner1d(a, b, out)
print(out)
```
Result:
```
[[ 1.00000000e+000  5.00000000e+000  9.00000000e+000]
 [ 6.91217735e-310 -1.45681599e+144 -1.45681599e+144]
 [-1.45681599e+144 -1.45681599e+144 -1.45681599e+144]
 [-1.45681599e+144 -1.45681599e+144 -1.45681599e+144]
 [-1.45681599e+144 -1.45681599e+144  6.32404027e-322]]
```
First row is the output, rest is garbage.

mhvk replied:
Hmm, that is not brilliant, indeed, rather seems like a bug to me! Checking with older versions (specifically, 1.17.4), though, I see that those had the same behaviour, so my sense would be to make this a separate issue, as this PR is rather large as it is already.

anirudh2290 replied:
sure, a seperate pr sounds good to me @mhvk .

mhvk replied:
@anirudh2290 - would you be willing to raise an issue so we do not forget?  Thanks!

seberg replied:
Agreed, its a problem. I think it may be problem we were discussing below? But there is probably more to it. This PR really only addresses the issue of non-existing axes (and generalize the approach to existing axes to also support `keepdims=True`).

To be clear, for `-1` (gufuncs) itself, the change is fairly small. In almost all cases `-1` currently does correctly mean "reduction (ok) axes". But, we also use it for unspecified output axes, which should normally be allocated using the iterator shape.
A clash of meanings which seems small for gufuncs, but was carefully worked around in reductions.

The PR resolves this, by marking all dimensions (-1 or not) if they are considered to be "reduced". And in that case, we know that the shape must be 1, even if we allocate the array in the iterator.
This is what makes the flagging approach so nice.  Not only does it mean we can push the "not too few output dimensions" check off into the iterator (which is all that this actually fixes for gufuncs). It also means the iterator can now allocate reduction results correctly even for arbitrarily complex reductions, and including `keepdims`.

mhvk commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -1890,22 +1903,34 @@ def test_reduce_noncontig_output(self):
-            assert_array_equal(check, f_reduce(a, axis=0, keepdims=True))
+            assert_array_equal(check, correct_out)
+
+    def test_reduce_output_no_subclass(self):
```
Maybe rename `test_reduce_output_subclass_ok` - since it seems that passing it in is fine, right?

mhvk reviewed (approved):
Triggered by the comments coming in, I finally had a look. I think this is *great* - it has become a very nice clean-up.  Apart from tiny comments, I wondered a bit whether `npyiter_get_op_axis` was doing too much double duty, i.e., whether it wouldn't be better to separate the question of whether something is a reduction axis from getting the real axis number out. But I ended up feeling it was fine as it was.

Again, really nice!

eric-wieser commented on numpy/core/src/multiarray/common.h:
```diff
@@ -342,5 +342,14 @@ NPY_NO_EXPORT PyArrayObject *
+ * probably. So we should consider making this public either as a macro or
+ * function (so that the way we flag the axis can be changed).
+ */
+#define NPY_ITER_REDUCTION_AXIS(axis) (axis + (NPY_MAX_INT >> 1))
```
Is this intended to be
```suggestion
#define NPY_ITER_REDUCTION_AXIS(axis) (axis + (1 << (NPY_BITSOF_INT - 2)))
```
seems weird to define this as `axis + 0x3FFF_FFFF` instead of `axis + 0x4000_0000`

seberg replied:
Since I cannot use bit-flags due to `-1`, just used a very large number instead... but sure changing only one bit for the positive case is probably nicer.

eric-wieser commented on numpy/core/src/multiarray/nditer_impl.h:
```diff
@@ -301,6 +301,40 @@ struct NpyIter_AD {
 /* Internal helper functions shared between implementation files */
+
+/**
+ * Undo the axis permutation of the iterator. When the operand has less
```
```suggestion
 * Undo the axis permutation of the iterator. When the operand has fewer
```

eric-wieser commented on numpy/core/src/multiarray/nditer_impl.h:
```diff
@@ -301,6 +301,40 @@ struct NpyIter_AD {
+ * @param ndim If `op_axes` is being used, this is the iterator dimension,
+ *             otherwise this is the operand dimension.
+ * @param perm he iterator axis permutation NIT_PERM(iter)
+ * @param axis_flipped Will be set to to true if this is a flipped axis and
```
What does `flipped` mean here? Reversed iteration order?

eric-wieser commented on numpy/core/src/multiarray/nditer_constr.c:
```diff
@@ -1410,6 +1410,38 @@ check_mask_for_writemasked_reduction(NpyIter *iter, int iop)
+ */
+static int
+npyiter_check_reduce_ok_and_set_flags(
+        NpyIter *iter, npy_uint32 flags, npyiter_opitflags *op_itflags) {
```
I think our style is usually closer to
```suggestion
        NpyIter *iter, npy_uint32 flags, npyiter_opitflags *op_itflags)
{
```

eric-wieser commented on numpy/core/src/multiarray/nditer_constr.c:
```diff
@@ -1410,6 +1410,38 @@ check_mask_for_writemasked_reduction(NpyIter *iter, int iop)
+    if (*op_itflags & NPY_OP_ITFLAG_WRITE) {
+        if (!(flags & NPY_ITER_REDUCE_OK)) {
+            PyErr_SetString(PyExc_ValueError,
+                    "output operand requires a reduction, but reduction is"
```
```suggestion
                    "output operand requires a reduction, but reduction is "
```

eric-wieser commented on numpy/core/tests/test_nditer.py:
```diff
@@ -1520,6 +1520,13 @@ def test_iter_allocate_output_errors():
+    assert_raises(ValueError, nditer, [a, None], ["reduce_ok"],
+                        [['readonly'], ['readwrite', 'allocate']],
+                        op_dtypes=[None, np.dtype('f4')],
+                        op_axes=[None, [0, np.newaxis, 2]])
```
What values here would not be an error?

seberg replied:
If the axes were 0 and 1 it would work. Hmm, you have to specify all the axes of the array here. I guess that is only true for allocated arrays probably.

eric-wieser reviewed (commented):
Don't have a good enough mental model of nditer to properly review all of this, but some of the commits are some nice cleanup, and will defer to @mhvk's call on this

eric-wieser commented on numpy/core/src/multiarray/einsum.c.src:
```diff
@@ -2740,7 +2740,7 @@ PyArray_EinsteinSum(char *subscripts, npy_intp nop,
     }
     for (idim = ndim_output; idim < ndim_iter; ++idim) {
-        op_axes[nop][idim] = -1;
+        op_axes[nop][idim] = NPY_ITER_REDUCTION_AXIS(-1);
```
Will this change make `np.einsum("...i,...i->...i", np.arange(3), np.array([1]))` an error? That is, should named dimensions in einsum be viewed as non-broadcastable?

seberg replied:
Yes, it changes that. I can undo. But einsum does not really have `keepdims` semantics nor is that notation einsum notation AFAIK.
Completely changing the resul behaviour by providing (a potentially wrong) output array seems undesirable to me?

eric-wieser replied:
What's the resulting error look like with this branch?

Not sure what you mean by that but behind einsum notation - that notation works for me

seberg replied:
Sorry, this is unchanged, I thought you had `np.einsum("...i->...i", np.arange(3), out=np.array([1]))`

If you do not provide an out array, no semantics are ever changed here.

seberg replied:
OK, sorry... I do not think anything changed at all. Since opposite broadcasting should be the only thing that changed (since only `np.newaxis` is marked here). But einsum specifically checks for `out` having too many dimensions.

seberg replied:
I.e. this is really just because I like this way better to specify the axes. The only thing where it makes a real difference currently would be reductions with `keepdims=True`, but there I think we may have had an additional check recently. Making the only actual bug fix/change in behaviour gufuncs...

seberg replied:
So what threw me off a bit, is that this does *not* address the fact that the output can cause normal expand the final result shape through broadcasting:
```
np.sum(arr, out=np.empty((3, 1)), axis=1, keepdims=True)
```
the output (writeable) array can modify the broadcast dimensions. That is actually a small regression here right now.

seberg replied:
For reductions, I can fix this trivially by using the no-broadcast flags on the input, which is obviously correct. slightly less nice error:
```
ValueError: non-broadcastable operand with shape (1,100) doesn't match the broadcast shape (3,100)
```
but it seems OK for me for now. For einsum its more complicated, since what we really want is the output to not have any say in the final shape, but there is no nice way to do that currently. However, there is also no change in behaviour here, so I think it does not matter for this PR.

mhvk replied:
That sounds good. But as a check, would one or the other solve the no-input-but-have-output case discussed in https://github.com/numpy/numpy/issues/15140?

mhvk replied:
OK, I had a quick look and this PR does not address #15140, so no worries about it!

seberg replied:
Hmmmm. I am wondering if we could mark an operand as output to indicate that it should not affect shape (or maybe using the WRITABLE flag is OK, but I fear that is a bit of a potential regression).

I suppose when there is no shape, the shape is just should be `()` (and if there are reduce, but keepdims, axis, they are always inserted with size 1 right now).

With your gufunc the question is whether you actually *want* the output to define the size in that case. Technically, that could be nice in some cases, unless you argue that the user should somehow provide `shape=...` in that case.
Although, for the iterator maybe that would be the same, and the gufunc machinery would just not pass the flag if there are no inputs?

seberg replied:
Oh, I am conflating two issues there. With no arguments we could already simply default to a 0-D iterator without much hassle probably...
How/if the output argument can affect the iterator shape is another issue.

mhvk replied:
Indeed, for my strange case I do not think the output shape has to affect the iterator; it was about automatically allocating gufunc dimensions if needed, even if there are no inputs.

seberg committed:
MAINT: Address review comments related to NPY_ITER_REDUCTION_AXIS

Incorporate suggestions by Marten and Eric.

seberg committed:
BUG: The reduction output must not cause the input to be broadcast

This was previously checked for, but during the refactor using
NPY_ITER_REDUCTION_AXIS to allocate the result in the iterator
instead of manually was lost here.
Einsum has a similar issue, but this is not modified here.

seberg commented:
OK, addressed the review comments and added a commit to fix the regression. The iterator is a confusing beast sometimes :(, but I hope this makes it at least slightly less confusing in the long run...

mhvk commented on numpy/core/tests/test_ufunc.py:
```diff
@@ -1922,7 +1922,19 @@ def test_reduce_wrong_dimension_output(self, f_reduce, keepdims, out_shape):
+        np.add.reduce(a, axis=-1, out=out_correct[:, 0], keepdims=False)
+        with assert_raises(ValueError):
+            np.add.reduce(a, axis=-1, out=out_incorrect, keepdims=True)
+            np.add.reduce(a, axis=-1, out=out_incorrect[:, 0], keepdims=False)
```
This needs a separate `with assert_raises`, otherwise it will not get executed.

seberg replied:
upst, fixed, thanks.

eric-wieser commented:
It's tempting to put the first four commits in their own PR, just to reduce the size of this one a little - depending on how close you think this is to ready, @mhvk.

mhvk commented:
@eric-wieser - My sense is that this is OK, but then I did review a simpler PR more focussed on just the actual change before (and already thought then that the approach is a very nice one, better than my #15142, though that may still be useful just to see that the actual change is not that large; the approach here is cleaner and done in a way that also allows the huge reduction in code for reductions). The added refactoring is a benefit - the nditer code is really rather opaque and I'm sure more could be done there but it is already quite an improvement.

I guess in the end it really depends whether you want to do a more thorough review yourself - if so, even though the commits are quite self-contained as is, I can see that it would be helpful for that to have the reorganization separated from the actual fix for broadcasting. But also important would be to ensure this is merged not close before a release, so there is ample time for it to be verified by projects that test against numpy-dev.

anirudh2290 commented on numpy/core/src/multiarray/nditer_constr.c:
```diff
@@ -1609,61 +1669,58 @@ npyiter_fill_axisdata(NpyIter *iter, npy_uint32 flags, npyiter_opitflags *op_itf
-                int i = axes[ndim-idim-1];
-                if (i >= 0) {
-                    if (bshape == 1 || op_cur == NULL) {
+                npy_bool reduction_axis;
```
nit: feel free to ignore. would be good to have some NPY_IT_DBG_PRINT in npyiter_fill_axis_data, since it hardly outputs anything if it touches this code path, except probably iter shape.

seberg commented:
Hmmm, I noticed the needs release notes tag that I had added. Does anyone think it is actually necessary? The most visible part would have been reductions, but we already fixed those. Matmul hacked around this IIRC (which I will check whether I can remove the hack). And other than that, I doubt we have any reduction-like gufuncs, so that this is all a very obvious bug-fix.

mhvk commented:
Good point. I think in the end it would be good to describe the change in the iterator C interface documentation (after all, we call the iterator differently now for gufuncs), but at some level this goes together with exposing `NPY_ITER_REDUCTION_AXIS`. So, perhaps it is fine to make that a separate issue? Right now, this does fix actual bugs!

seberg commented:
Yeah, I would like to expose this, just need to get both the C- and python API right. The main thing I would like to be sure is that the way we set the flag is well expandable in the future, just in case we want to add more flags for some reason.
As such I think this API is much better than "reduce ok"...

anirudh2290 reviewed (approved):
LGTM! Mostly checked only a part of this PR related to https://github.com/numpy/numpy/issues/15139

mattip commented:
Does this impact benchmarks at all?

seberg commented:
This can only affect tiny overheads, and I am unwilling to judge those. There are *far* bigger fish on the overhead side. IIRC normal reductions were microscopically slower, while `keepdims=True` was faster, tiny spill-over to other things is of course possible.
It may be that part of the refactor was actually because it removed overheads larger than anything added by the flag checking...

seberg commented:
Well, I ran the ufunc and reduction benchmarks and of course no changes (they do not test overheads anyway).

From me this is finished. I spend a lot of effort in making each commit self contained, if necessary I can split it up, but until someone ask me to do that I have to assume that the self contained commits are OK for now?

mattip commented:
Thanks @seberg

mhvk commented:
Yay! Very nice!
